### PR TITLE
requirements_common.txt: Upgrade to web.py==0.40

### PR DESCRIPTION
**Pull request by cclauss**
https://github.com/webpy/webpy/blob/master/ChangeLog.txt

Blocked by ~internetarchive/infogami#46~

Fails on:
```
infogami/utils/view.py:69: in <module>
    utf8=web.utf8,
E   AttributeError: 'module' object has no attribute 'utf8'
```
https://travis-ci.org/internetarchive/openlibrary/jobs/590405210#L773

### CODE REVIEW

**Comment by iredmail** [comment]
@cclauss I did a quick check, `web.utf8` was removed on Jun 24, 2016 by @Zashas:
https://github.com/webpy/webpy/commit/99400202e6f3fcf60957cbd1ecc2db7d884c8dfa#diff-23d7a79688f6a6dbb55d915adaadb27aL372

You can fix it by simply replacing `utf8=web.utf8,` by `utf8=web.safestr,` in infogami/utils/view.py (line 69).

**Comment by cclauss** [comment]
Blocked by internetarchive/infogami#46

**Comment by cclauss** [comment]
Python 2 blocked by:
```
ImportError while loading conftest '/home/travis/build/internetarchive/openlibrary/openlibrary/conftest.py'.
../../../virtualenv/python2.7.15/lib/python2.7/site-packages/six.py:709: in exec_
    exec("""exec _code_ in _globs_, _locs_""")
openlibrary/conftest.py:18: in <module>
    from openlibrary.mocks.mock_ol import ol
openlibrary/mocks/mock_ol.py:37: in <module>
    class OLBrowser(web.AppBrowser):
E   AttributeError: 'module' object has no attribute 'AppBrowser'
```

Python 3 blocked by internetarchive/infogami#45

**Comment by iredmail** [comment]
`AppBrowser` is not imported to `web` namespace, use code like below instead:

```
from web.browser import AppBrowser
```

**Comment by cclauss** [comment]
Python 2 is blocked by
```
ImportError while importing test module '/home/travis/build/internetarchive/openlibrary/openlibrary/coverstore/tests/test_doctests.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../../virtualenv/python2.7.15/lib/python2.7/site-packages/six.py:709: in exec_
    exec("""exec _code_ in _globs_, _locs_""")
openlibrary/coverstore/tests/test_doctests.py:2: in <module>
    from infogami.infobase.tests.test_doctests import find_doctests, run_doctest
E   ImportError: cannot import name find_doctests
```

Python 3 is blocked by internetarchive/infogami#45 or internetarchive/infogami#47

**Comment by cdrini** [comment]
It looks like this commit removed the function that's being imported here: https://github.com/internetarchive/infogami/commit/384d6c9d4c216f0f9c44c664b399ca558404f438#diff-d3bca5f4a6030a1c66a821667c6abeaa (cc: @hornc )

**Comment by cclauss** [comment]
That change is nowhere in the diff.  I have never seen something like that!  Closing.  I will create from scratch on a clean branch.

**Comment by cdrini** [comment]
Oh, that diff's not part of this PR! This PR looks good :) That commit is already in master; it must have been merged a while back.

File: .travis.yml [hunk 1]
```diff
@@ -19,17 +19,16 @@ install:
...
+  # refresh the git submodule ./vendor/infogami
+  # git commit those changes into this pull request
   # if Travis is running Python 3, then
-  #   refresh the git submodule ./vendor/acs4_py
-  #   refresh the git submodule ./vendor/infogami
-  #   git commit those changes into this pull request
-  #   TODO: flake8 ./vendor --select=E9,F63,F7,F82 to unsure new versions pass
   #   create a legacy Python for npm/node until nodejs/node#25789 is resolved
+  - pushd vendor/acs4_py  && git pull origin master && popd
+  - pushd vendor/infogami && git pull origin master && popd
```
Comment (hornc) [inline .travis.yml]: I'm not sure about this method to update the required versions of our submodules for python3. It looks like this is only happening on Travis so there is no mechanism to checkout the correct code in any other environment. The code SHA/tag is not fixed either, like it is with the standard gitmodules in `vendor` https://github.com/internetarchive/openlibrary/tree/master/vendor

Is there a way we can leverage gitmodules to check out two different points in infogami and acs4 depending on the Python version that applies in all cases, using the existing gitmodules mechanism? (I have not begun to investigate this yet)

Comment (cclauss) [reply .travis.yml]: acs4_py can be removed.  It is already 💯 ready for Python 3.

**Comment by cclauss** [comment]
How do we get the __testing__ done?

File: openlibrary/mocks/mock_ol.py [hunk 2]
```diff
@@ -1,11 +1,16 @@
 import os
-import pytest
 import re
+
+import pytest
 import web
+try:
+    from web.browser import AppBrowser
+except ImportError:
```
Comment (tfmorris) [inline openlibrary/mocks/mock_ol.py]: I don't think that the old form of the import is needed. I guess it doesn't hurt anything, but is there a reason to include it given that we don't support both web.py 0.39 and 0.40?

File: openlibrary/plugins/admin/code.py [hunk 3]
```diff
@@ -15,10 +15,10 @@
 from infogami.utils import delegate
 from infogami.utils.view import render, public
 from infogami.utils.context import context
-from infogami.utils.view import add_flash_message
+from infogami.utils.flash import add_flash_message
```
Comment (tfmorris) [inline openlibrary/plugins/admin/code.py]: The incompatible change to infogami has been reverted, so these imports can be put back the way they were.

Comment (tfmorris) [reply openlibrary/plugins/admin/code.py]: I just pushed a commit that reverts the 7 commits that made these changes, so hopefully this is all set.

**Comment by tfmorris** [comment]
@hornc In https://github.com/internetarchive/openlibrary/pull/2897#pullrequestreview-345616773, you said:

> Moving from web.py 0.39 to 0.40 is going to take quite a bit of planning and discussion to get across the line, ...

Can you expand on this and outline what you think is required from a planning and discussion point of view? Obviously it needs to be put on dev for testing, but what else is needed?

**Comment by tfmorris** [comment]
I've rebased this branch and squashed all the commits that cancel each other out. In the process, I also restored https://github.com/internetarchive/openlibrary/pull/2416/commits/30cb722fec6178b8861d3de27d85c733d631032f which got accidentally overwritten in the last merge from master.

Hopefully it'll be easier for folks to review the cleaned up branch/PR.

**Comment by cclauss** [comment]
Closing in favor of #3027

**Pull request closed without merging**